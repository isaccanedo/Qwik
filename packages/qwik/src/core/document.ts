/**
 * Contains additional items which Qwik patches on the Document.
 */
export interface QwikDocument extends Document {
  __q_context__?: [HTMLElement, Event, URL] | {};
}

export const _is_module = true;
