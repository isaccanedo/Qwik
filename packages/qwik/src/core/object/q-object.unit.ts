import {
  createSubscriptionManager,
  getOrCreateProxy,
  QObjectRecursive,
  unwrapProxy,
} from './q-object';

describe('q-object', () => {
  beforeEach(() => {});
  const map: any = {
    $subsManager$: createSubscriptionManager(),
    $proxyMap$: new WeakMap(),
  };

  it('should create QObject', () => {
    const obj = getOrCreateProxy({ salutation: 'Hello', name: 'World' }, map);
    expect(obj).toEqual({ salutation: 'Hello', name: 'World' });
  });

  describe('read write proxy', () => {
    it('should support basic operations', () => {
      const value = { a: 1, b: 2 };
      const proxy = getOrCreateProxy(value, map);
      expect(proxy.a).toBe(1);
      expect(proxy.b).toBe(2);
      expect(unwrapProxy(proxy as any)).toBe(value);
      expect(++proxy.b).toBe(3);
      expect('b' in proxy).toBe(true);
      expect(Object.keys(proxy)).toEqual(Object.keys(value));
    });

    it('should support child objects', () => {
      const child = { a: 1, b: 2 };
      const parent = { child: child };
      const proxy = getOrCreateProxy(parent, map, QObjectRecursive);
      expect(proxy.child.a).toBe(1);
      const pChild = proxy.child;
      expect(proxy.child).not.toBe(child);
      proxy.child = pChild;
      const child2 = {};
      proxy.child = child2 as any;
      expect(parent.child).toBe(child2);
    });

    describe('array', () => {
      it('should support arrays', () => {
        const child = { a: 'a' };
        const list = [1, child];
        const pList = getOrCreateProxy(list, map);
        expect(Object.keys(pList)).toEqual(Object.keys(list));
        expect(pList).toEqual(list);
        const copy = [] as any;
        for (const key in pList) {
          if (Object.prototype.hasOwnProperty.call(pList, key)) {
            copy[key] = pList[key];
          }
        }
        copy.length = 0;
        for (const v of pList) {
          copy.push(v);
        }
        expect(copy).toEqual(list);

        pList.push(2, child);
        expect(pList).toEqual([1, child, 2, child]);
      });
    });

    it('should support equality', () => {
      // TODO(misko): I don't think it is possible with proxy.
    });
  });
});
