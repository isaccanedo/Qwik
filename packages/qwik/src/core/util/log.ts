import { isElement } from '../../testing/html';
import { tryGetContext } from '../props/props';
import { QHostAttr } from './markers';
import { qDev } from './qdev';

const STYLE = qDev
  ? `background: #564CE0; color: white; padding: 2px 3px; border-radius: 2px; font-size: 0.8em;`
  : '';

export const logError = (message?: any, ...optionalParams: any[]) => {
  const err = message instanceof Error ? message : new Error(message);
  // eslint-disable-next-line no-console
  console.error('%cQWIK ERROR', STYLE, err, ...printParams(optionalParams));
  return err;
};

export const logErrorAndStop = (message?: any, ...optionalParams: any[]) => {
  logError(message, ...optionalParams);
  // eslint-disable-next-line no-debugger
  debugger;
};

export const logWarn = (message?: any, ...optionalParams: any[]) => {
  // eslint-disable-next-line no-console
  if (qDev) {
    console.warn('%cQWIK WARN', STYLE, message, ...printParams(optionalParams));
  }
};

export const logDebug = (message?: string, ...optionalParams: any[]) => {
  if (qDev) {
    // eslint-disable-next-line no-console
    console.debug('%cQWIK', STYLE, message, ...printParams(optionalParams));
  }
};

const printParams = (optionalParams: any[]) => {
  if (qDev) {
    return optionalParams.map((p) => {
      if (isElement(p)) {
        return printElement(p);
      }
      return p;
    });
  }
  return optionalParams;
};

const printElement = (el: Element) => {
  const ctx = tryGetContext(el);
  const isComponent = el.hasAttribute(QHostAttr);
  const isServer: boolean = /*#__PURE__*/ (() =>
    typeof process !== 'undefined' && !!process.versions && !!process.versions.node)();

  return {
    isComponent,
    tagName: el.tagName,
    renderQRL: ctx?.$renderQrl$?.getSymbol(),
    element: isServer ? undefined : el,
    ctx: isServer ? undefined : ctx,
  };
};
