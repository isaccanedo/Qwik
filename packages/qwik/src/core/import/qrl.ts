import { EMPTY_ARRAY } from '../util/flyweight';
import type { QRL } from './qrl.public';
import { assertQrl, createQrl, QRLInternal } from './qrl-class';
import { isFunction, isString } from '../util/types';
import type { CorePlatform } from '../platform/types';
import { getDocument } from '../util/dom';
import { logError } from '../util/log';
import { tryGetInvokeContext } from '../use/use-core';
import {
  codeToText,
  qError,
  QError_dynamicImportFailed,
  QError_runtimeQrlNoElement,
  QError_unknownTypeArgument,
} from '../error/error';

let runtimeSymbolId = 0;
const RUNTIME_QRL = '/runtimeQRL';
const INLINED_QRL = '/inlinedQRL';

// https://regexr.com/68v72
const EXTRACT_IMPORT_PATH = /\(\s*(['"])([^\1]+)\1\s*\)/;

// https://regexr.com/690ds
const EXTRACT_SELF_IMPORT = /Promise\s*\.\s*resolve/;

// https://regexr.com/6a83h
const EXTRACT_FILE_NAME = /[\\/(]([\w\d.\-_]+\.(js|ts)x?):/;

// <docs markdown="../readme.md#qrl">
// !!DO NOT EDIT THIS COMMENT DIRECTLY!!!
// (edit ../readme.md#qrl instead)
/**
 * Used by Qwik Optimizer to point to lazy-loaded resources.
 *
 * This function should be used by the Qwik Optimizer only. The function should not be directly
 * referred to in the source code of the application.
 *
 * @see `QRL`, `$(...)`
 *
 * @param chunkOrFn - Chunk name (or function which is stringified to extract chunk name)
 * @param symbol - Symbol to lazy load
 * @param lexicalScopeCapture - a set of lexically scoped variables to capture.
 * @alpha
 */
// </docs>
export const qrl = <T = any>(
  chunkOrFn: string | (() => Promise<any>),
  symbol: string,
  lexicalScopeCapture: any[] = EMPTY_ARRAY
): QRL<T> => {
  let chunk: string;
  let symbolFn: null | (() => Promise<Record<string, any>>) = null;
  if (isString(chunkOrFn)) {
    chunk = chunkOrFn;
  } else if (isFunction(chunkOrFn)) {
    symbolFn = chunkOrFn;
    let match: RegExpMatchArray | null;
    const srcCode = String(chunkOrFn);
    if ((match = srcCode.match(EXTRACT_IMPORT_PATH)) && match[2]) {
      chunk = match[2];
    } else if ((match = srcCode.match(EXTRACT_SELF_IMPORT))) {
      const ref = 'QWIK-SELF';
      const frames = new Error(ref).stack!.split('\n');
      const start = frames.findIndex((f) => f.includes(ref));
      const frame = frames[start + 2];
      match = frame.match(EXTRACT_FILE_NAME);
      if (!match) {
        chunk = 'main';
      } else {
        chunk = match[1];
      }
    } else {
      throw qError(QError_dynamicImportFailed, srcCode);
    }
  } else {
    throw qError(QError_unknownTypeArgument, chunkOrFn);
  }

  // Unwrap subscribers
  const qrl = createQrl<T>(chunk, symbol, null, symbolFn, null, lexicalScopeCapture, null);
  const ctx = tryGetInvokeContext();
  if (ctx && ctx.$element$) {
    qrl.$setContainer$(ctx.$element$);
  }
  return qrl;
};

export const runtimeQrl = <T>(
  symbol: T,
  lexicalScopeCapture: any[] = EMPTY_ARRAY
): QRLInternal<T> => {
  return createQrl<T>(
    RUNTIME_QRL,
    's' + runtimeSymbolId++,
    symbol,
    null,
    null,
    lexicalScopeCapture,
    null
  );
};

/**
 * @alpha
 */
export const inlinedQrl = <T>(
  symbol: T,
  symbolName: string,
  lexicalScopeCapture: any[] = EMPTY_ARRAY
): QRL<T> => {
  // Unwrap subscribers
  return createQrl<T>(INLINED_QRL, symbolName, symbol, null, null, lexicalScopeCapture, null);
};

export interface QRLSerializeOptions {
  $platform$?: CorePlatform;
  $element$?: Element;
  $getObjId$?: (obj: any) => string | null;
}

export const stringifyQRL = (qrl: QRLInternal, opts: QRLSerializeOptions = {}) => {
  assertQrl(qrl);
  let symbol = qrl.$symbol$;
  let chunk = qrl.$chunk$;
  const refSymbol = qrl.$refSymbol$ ?? symbol;
  const platform = opts.$platform$;
  const element = opts.$element$;
  if (platform) {
    const result = platform.chunkForSymbol(refSymbol);
    if (result) {
      chunk = result[1];
      if (!qrl.$refSymbol$) {
        symbol = result[0];
      }
    }
  }
  const parts: string[] = [chunk];
  if (symbol && symbol !== 'default') {
    parts.push('#', symbol);
  }
  const capture = qrl.$capture$;
  const captureRef = qrl.$captureRef$;
  if (opts.$getObjId$) {
    if (captureRef && captureRef.length) {
      const capture = captureRef.map(opts.$getObjId$);
      parts.push(`[${capture.join(' ')}]`);
    }
  } else if (capture && capture.length > 0) {
    parts.push(`[${capture.join(' ')}]`);
  }
  const qrlString = parts.join('');
  if (qrl.$chunk$ === RUNTIME_QRL && element) {
    const qrls: Set<QRL> = (element as any).__qrls__ || ((element as any).__qrls__ = new Set());
    qrls.add(qrl);
  }
  return qrlString;
};

export const qrlToUrl = (element: Element, qrl: QRLInternal): URL => {
  return new URL(stringifyQRL(qrl), getDocument(element).baseURI);
};

/**
 * `./chunk#symbol[captures]
 */
export const parseQRL = (qrl: string, el?: Element): QRLInternal => {
  const endIdx = qrl.length;
  const hashIdx = indexOf(qrl, 0, '#');
  const captureIdx = indexOf(qrl, hashIdx, '[');

  const chunkEndIdx = Math.min(hashIdx, captureIdx);
  const chunk = qrl.substring(0, chunkEndIdx);

  const symbolStartIdx = hashIdx == endIdx ? hashIdx : hashIdx + 1;
  const symbolEndIdx = captureIdx;
  const symbol =
    symbolStartIdx == symbolEndIdx ? 'default' : qrl.substring(symbolStartIdx, symbolEndIdx);

  const captureStartIdx = captureIdx;
  const captureEndIdx = endIdx;
  const capture =
    captureStartIdx === captureEndIdx
      ? EMPTY_ARRAY
      : qrl.substring(captureStartIdx + 1, captureEndIdx - 1).split(' ');

  if (chunk === RUNTIME_QRL) {
    logError(codeToText(QError_runtimeQrlNoElement), qrl);
  }
  const iQrl = createQrl<any>(chunk, symbol, null, null, capture, null, null);
  if (el) {
    iQrl.$setContainer$(el);
  }
  return iQrl;
};

const indexOf = (text: string, startIdx: number, char: string) => {
  const endIdx = text.length;
  const charIdx = text.indexOf(char, startIdx == endIdx ? 0 : startIdx);
  return charIdx == -1 ? endIdx : charIdx;
};
