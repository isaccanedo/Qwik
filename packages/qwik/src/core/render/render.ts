import { getContext } from '../props/props';
import { isArray, ValueOrPromise } from '../util/types';
import { RenderContext, smartUpdateChildren, updateProperties } from './cursor';
import type { ProcessedJSXNode } from './jsx/types/jsx-node';
import { HOST_TYPE } from './jsx/jsx-runtime';
export type ComponentRenderQueue = Promise<HTMLElement[]>[];

export const visitJsxNode = (
  ctx: RenderContext,
  elm: Element,
  jsxNode: ProcessedJSXNode | ProcessedJSXNode[] | undefined,
  isSvg: boolean
): ValueOrPromise<void> => {
  if (jsxNode === undefined) {
    return smartUpdateChildren(ctx, elm, [], 'root', isSvg);
  }
  if (isArray(jsxNode)) {
    return smartUpdateChildren(ctx, elm, jsxNode.flat(), 'root', isSvg);
  } else if (jsxNode.$type$ === HOST_TYPE) {
    updateProperties(ctx, getContext(elm), jsxNode.$props$, isSvg);
    return smartUpdateChildren(ctx, elm, jsxNode.$children$ || [], 'root', isSvg);
  } else {
    return smartUpdateChildren(ctx, elm, [jsxNode], 'root', isSvg);
  }
};
