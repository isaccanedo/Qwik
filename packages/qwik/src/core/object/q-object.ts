import { assertEqual } from '../assert/assert';
import {
  qError,
  QError_immutableProps,
  QError_onlyLiteralWrapped,
  QError_onlyObjectWrapped,
  QError_verifySerializable,
} from '../error/error';
import { isQrl } from '../import/qrl-class';
import { ContainerState, notifyChange } from '../render/notify-render';
import { tryGetInvokeContext } from '../use/use-core';
import { isDocument, isElement, isNode } from '../util/element';
import { logWarn } from '../util/log';
import { qDev } from '../util/qdev';
import type { Subscriber } from '../use/use-subscriber';
import { tryGetContext } from '../props/props';
import { RenderEvent } from '../util/markers';
import { isArray, isFunction, isObject } from '../util/types';
import { isPromise } from '../util/promises';

export type ObjToProxyMap = WeakMap<any, any>;
export type QObject<T extends {}> = T & { __brand__: 'QObject' };

export const QObjectRecursive = 1 << 0;
export const QObjectImmutable = 1 << 1;

/**
 * Creates a proxy that notifies of any writes.
 */
export const getOrCreateProxy = <T extends object>(
  target: T,
  containerState: ContainerState,
  flags = 0
): T => {
  const proxy = containerState.$proxyMap$.get(target);
  if (proxy) {
    return proxy;
  }
  return createProxy(target, containerState, flags, undefined);
};

export const createProxy = <T extends object>(
  target: T,
  containerState: ContainerState,
  flags: number,
  subs?: Map<Element, Set<string>>
): T => {
  assertEqual(unwrapProxy(target), target, 'Unexpected proxy at this location');
  assertEqual(containerState.$proxyMap$.has(target), false, 'Proxy was already created');

  if (!isObject(target)) {
    throw qError(QError_onlyObjectWrapped, target);
  }
  if (target.constructor !== Object && !isArray(target)) {
    throw qError(QError_onlyLiteralWrapped, target);
  }

  const manager = containerState.$subsManager$.$getLocal$(target, subs);
  const proxy = new Proxy(
    target,
    new ReadWriteProxyHandler(containerState, manager, flags)
  ) as any as T;
  containerState.$proxyMap$.set(target, proxy);
  return proxy;
};

const QOjectTargetSymbol = Symbol();
const QOjectFlagsSymbol = Symbol();

export type TargetType = Record<string | symbol, any>;

export type SubscriberMap = Map<Subscriber, Set<string> | null>;
export type ObjToSubscriberMap = WeakMap<any, LocalSubscriptionManager>;
export type SubscriberToSubscriberMap = Map<Subscriber, Set<SubscriberMap>>;

export interface SubscriptionManager {
  $tryGetLocal$(obj: any): LocalSubscriptionManager | undefined;
  $getLocal$(obj: any, map?: SubscriberMap): LocalSubscriptionManager;
  $clearSub$: (sub: Subscriber) => void;
}

export interface LocalSubscriptionManager {
  $subs$: SubscriberMap;
  $notifySubs$: (key?: string | undefined) => void;
  $addSub$: (subscriber: Subscriber, key?: string) => void;
}

export const createSubscriptionManager = (): SubscriptionManager => {
  const objToSubs: ObjToSubscriberMap = new Map();
  const subsToObjs: SubscriberToSubscriberMap = new Map();

  const clearSub = (sub: Subscriber) => {
    const subs = subsToObjs.get(sub);
    if (subs) {
      subs.forEach((s) => {
        s.delete(sub);
      });
      subsToObjs.delete(sub);
      subs.clear();
    }
  };

  const tryGetLocal = (obj: any) => {
    assertEqual(getProxyTarget(obj), undefined, 'object can not be be a proxy');
    return objToSubs.get(obj);
  };

  const trackSubToObj = (subscriber: Subscriber, map: SubscriberMap) => {
    let set = subsToObjs.get(subscriber);
    if (!set) {
      subsToObjs.set(subscriber, (set = new Set()));
    }
    set.add(map);
  };

  const getLocal = (obj: any, initialMap?: SubscriberMap) => {
    let local = tryGetLocal(obj);
    if (local) {
      assertEqual(initialMap, undefined, 'subscription map can not be set to an existing object');
    } else {
      const map = !initialMap ? (new Map() as SubscriberMap) : initialMap;
      map.forEach((_, key) => {
        trackSubToObj(key, map);
      });
      objToSubs.set(
        obj,
        (local = {
          $subs$: map,
          $addSub$(subscriber: Subscriber, key?: string) {
            if (key == null) {
              map.set(subscriber, null);
            } else {
              let sub = map.get(subscriber);
              if (sub === undefined) {
                map.set(subscriber, (sub = new Set()));
              }
              if (sub) {
                sub.add(key);
              }
            }
            trackSubToObj(subscriber, map);
          },
          $notifySubs$(key?: string) {
            map.forEach((value, subscriber) => {
              if (value === null || !key || value.has(key)) {
                notifyChange(subscriber);
              }
            });
          },
        })
      );
    }
    return local;
  };

  return {
    $tryGetLocal$: tryGetLocal,
    $getLocal$: getLocal,
    $clearSub$: clearSub,
  };
};

class ReadWriteProxyHandler implements ProxyHandler<TargetType> {
  constructor(
    private $containerState$: ContainerState,
    private $manager$: LocalSubscriptionManager,
    private $flags$: number
  ) {}

  get(target: TargetType, prop: string | symbol): any {
    if (typeof prop === 'symbol') {
      if (prop === QOjectTargetSymbol) return target;
      if (prop === QOjectFlagsSymbol) return this.$flags$;
      return target[prop];
    }
    let subscriber: Subscriber | undefined | null;
    const invokeCtx = tryGetInvokeContext();
    const recursive = (this.$flags$ & QObjectRecursive) !== 0;
    const immutable = (this.$flags$ & QObjectImmutable) !== 0;
    if (invokeCtx) {
      subscriber = invokeCtx.$subscriber$;
    }
    let value = target[prop];
    if (isMutable(value)) {
      value = value.v;
    } else if (immutable) {
      subscriber = null;
    }
    if (subscriber) {
      const isA = isArray(target);
      this.$manager$.$addSub$(subscriber, isA ? undefined : prop);
    }
    return recursive ? wrap(value, this.$containerState$) : value;
  }

  set(target: TargetType, prop: string | symbol, newValue: any): boolean {
    if (typeof prop === 'symbol') {
      target[prop] = newValue;
      return true;
    }
    const immutable = (this.$flags$ & QObjectImmutable) !== 0;
    if (immutable) {
      throw qError(QError_immutableProps);
    }
    const recursive = (this.$flags$ & QObjectRecursive) !== 0;
    const unwrappedNewValue = recursive ? unwrapProxy(newValue) : newValue;
    if (qDev) {
      verifySerializable(unwrappedNewValue);
      const invokeCtx = tryGetInvokeContext();
      if (invokeCtx && invokeCtx.$event$ === RenderEvent) {
        logWarn(
          'State mutation inside render function. Move mutation to useWatch(), useClientEffect() or useServerMount()',
          invokeCtx.$hostElement$,
          prop
        );
      }
    }
    const isA = isArray(target);
    if (isA) {
      target[prop as any] = unwrappedNewValue;
      this.$manager$.$notifySubs$();
      return true;
    }

    const oldValue = target[prop];
    if (oldValue !== unwrappedNewValue) {
      target[prop] = unwrappedNewValue;
      this.$manager$.$notifySubs$(prop);
    }
    return true;
  }

  has(target: TargetType, property: string | symbol) {
    if (property === QOjectTargetSymbol) return true;
    if (property === QOjectFlagsSymbol) return true;

    return Object.prototype.hasOwnProperty.call(target, property);
  }

  ownKeys(target: TargetType): ArrayLike<string | symbol> {
    let subscriber: Subscriber | null | undefined = null;
    const invokeCtx = tryGetInvokeContext();
    if (invokeCtx) {
      subscriber = invokeCtx.$subscriber$;
    }
    if (subscriber) {
      this.$manager$.$addSub$(subscriber);
    }
    return Object.getOwnPropertyNames(target);
  }
}

const wrap = <T>(value: T, containerState: ContainerState): T => {
  if (isQrl(value)) {
    return value;
  }
  if (isObject(value)) {
    if (Object.isFrozen(value)) {
      return value;
    }
    const nakedValue = unwrapProxy(value);
    if (nakedValue !== value) {
      // already a proxy return;
      return value;
    }
    if (isNode(nakedValue)) {
      return value;
    }
    if (!shouldSerialize(nakedValue)) {
      return value;
    }
    if (qDev) {
      verifySerializable<T>(value);
    }
    const proxy = containerState.$proxyMap$.get(value);
    return proxy ? proxy : getOrCreateProxy(value as any, containerState, QObjectRecursive);
  } else {
    return value;
  }
};

export const verifySerializable = <T>(value: T): T => {
  const seen = new Set();
  return _verifySerializable(value, seen);
};

const _verifySerializable = <T>(value: T, seen: Set<any>): T => {
  const unwrapped = unwrapProxy(value);
  if (unwrapped == null) {
    return value;
  }
  if (shouldSerialize(unwrapped)) {
    if (seen.has(unwrapped)) {
      return value;
    }
    seen.add(unwrapped);
    if (isQrl(unwrapped)) {
      return value;
    }
    switch (typeof unwrapped) {
      case 'object':
        if (isArray(unwrapped)) {
          for (const item of unwrapped) {
            _verifySerializable(item, seen);
          }
          return value;
        }
        if (Object.getPrototypeOf(unwrapped) === Object.prototype) {
          for (const item of Object.values(unwrapped)) {
            _verifySerializable(item, seen);
          }
          return value;
        }
        if (isPromise(unwrapped)) return value;
        if (isElement(unwrapped)) return value;
        if (isDocument(unwrapped)) return value;
        break;
      case 'boolean':
      case 'string':
      case 'number':
        return value;
    }
    throw qError(QError_verifySerializable, unwrapped);
  }
  return value;
};
const noSerializeSet = /*#__PURE__*/ new WeakSet<any>();

export const shouldSerialize = (obj: any): boolean => {
  if (isObject(obj) || isFunction(obj)) {
    return !noSerializeSet.has(obj);
  }
  return true;
};

/**
 * @alpha
 */
export type NoSerialize<T> = (T & { __no_serialize__: true }) | undefined;

// <docs markdown="../readme.md#noSerialize">
// !!DO NOT EDIT THIS COMMENT DIRECTLY!!!
// (edit ../readme.md#noSerialize instead)
/**
 * Marks a property on a store as non-serializable.
 *
 * At times it is necessary to store values on a store that are non-serializable. Normally this
 * is a runtime error as Store wants to eagerly report when a non-serializable property is
 * assigned to it.
 *
 * You can use `noSerialize()` to mark a value as non-serializable. The value is persisted in the
 * Store but does not survive serialization. The implication is that when your application is
 * resumed, the value of this object will be `undefined`. You will be responsible for recovering
 * from this.
 *
 * See: [noSerialize Tutorial](http://qwik.builder.io/tutorial/store/no-serialize)
 *
 * @alpha
 */
// </docs>
export const noSerialize = <T extends {}>(input: T): NoSerialize<T> => {
  noSerializeSet.add(input);
  return input as any;
};

// <docs markdown="../readme.md#immutable">
// !!DO NOT EDIT THIS COMMENT DIRECTLY!!!
// (edit ../readme.md#immutable instead)
/**
 * Mark an object as immutable, preventing Qwik from creating subscriptions on that object.
 *
 * Qwik automatically creates subscriptions on store objects created by `useStore()`. By marking
 * an object as `immutable`, it hints to Qwik that the properties of this object will not change,
 * and therefore there is no need to create subscriptions for those objects.
 *
 * @alpha
 */
// </docs>
export const immutable = <T extends {}>(input: T): Readonly<T> => {
  return Object.freeze(input);
};

// <docs markdown="../readme.md#mutable">
// !!DO NOT EDIT THIS COMMENT DIRECTLY!!!
// (edit ../readme.md#mutable instead)
/**
 * Mark property as mutable.
 *
 * Qwik assumes that all bindings in components are immutable by default. This is done for two
 * reasons:
 *
 * 1. JSX does not allow Qwik runtime to know if a binding is static or mutable.
 *    `<Example valueA={123} valueB={exp}>` At runtime there is no way to know if `valueA` is
 * immutable.
 * 2. If Qwik assumes that properties are immutable, then it can do a better job data-shaking the
 * amount of code that needs to be serialized to the client.
 *
 * Because Qwik assumes that bindings are immutable by default, it needs a way for a developer to
 * let it know that binding is mutable. `mutable()` function serves that purpose.
 * `<Example valueA={123} valueB={mutable(exp)}>`. In this case, the Qwik runtime can correctly
 * recognize that the `Example` props are mutable and need to be serialized.
 *
 * See: [Mutable Props Tutorial](http://qwik.builder.io/tutorial/props/mutable) for an example
 *
 * @alpha
 */
// </docs>
export const mutable = <T>(v: T): MutableWrapper<T> => {
  return {
    [MUTABLE]: true,
    v,
  };
};

export const isConnected = (sub: Subscriber): boolean => {
  if (isElement(sub)) {
    return !!tryGetContext(sub) || sub.isConnected;
  } else {
    return isConnected(sub.el);
  }
};

const MUTABLE = Symbol('mutable');

// <docs markdown="../readme.md#MutableWrapper">
// !!DO NOT EDIT THIS COMMENT DIRECTLY!!!
// (edit ../readme.md#MutableWrapper instead)
/**
 * A marker object returned by `mutable()` to identify that the binding is mutable.
 *
 * @alpha
 */
// </docs>
export interface MutableWrapper<T> {
  /**
   * A marker symbol.
   */
  [MUTABLE]: true;
  /**
   * Mutable value.
   */
  v: T;
}

export const isMutable = (v: any): v is MutableWrapper<any> => {
  return isObject(v) && v[MUTABLE] === true;
};

/**
 * @alpha
 */
export const unwrapProxy = <T>(proxy: T): T => {
  return getProxyTarget<T>(proxy) ?? proxy;
};

export const getProxyTarget = <T = Record<string, any>>(obj: T): T | undefined => {
  if (isObject(obj)) {
    return (obj as any)[QOjectTargetSymbol];
  }
  return undefined;
};

export const getProxyFlags = <T = Record<string, any>>(obj: T): number | undefined => {
  if (isObject(obj)) {
    return (obj as any)[QOjectFlagsSymbol];
  }
  return undefined;
};
