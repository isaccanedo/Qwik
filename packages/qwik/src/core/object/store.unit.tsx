import { createDocument } from '../../testing/document';
import { useStore } from '../use/use-store.public';
import { render } from '../render/render.public';
import { getQwikJSON } from './store';
import { useLexicalScope } from '../use/use-lexical-scope.public';
import { component$ } from '../component/component.public';
import { noSerialize } from './q-object';
import { $ } from '../import/qrl.public';
import { logDebug } from '../util/log';
import { runtimeQrl } from '../import/qrl';
import { pauseContainer } from '../object/store';
import { useDocument } from '../use/use-document.public';
import { useHostElement } from '../use/use-host-element.public';

describe('store', () => {
  let document: Document;
  let div: HTMLElement;

  beforeEach(() => {
    document = createDocument();
    div = document.createElement('div');
    document.body.appendChild(div);
  });

  it('should serialize content', async () => {
    await render(
      document.body,
      <div>
        <LexicalScope />
      </div>
    );
    await pauseContainer(document.body);
    const script = getQwikJSON(document.body)!;
    expect(JSON.parse(script.textContent!)).toMatchSnapshot();
  });
});

export const LexicalScope_render = () => {
  const [a, b, c, d, e, f, g, h, state, noserialize] = useLexicalScope();
  return (
    <section>
      <p>{JSON.stringify(a)}</p>
      <p>{JSON.stringify(b)}</p>
      <p>{JSON.stringify(c)}</p>
      <p>{String(d)}</p>
      <p>{String(e)}</p>
      <p>{JSON.stringify(f)}</p>
      <p>{JSON.stringify(g)}</p>
      <p>{JSON.stringify(h)}</p>
      <p>{noserialize.text}</p>
      <button onDocumentClick$={() => state.count++}>Rerender {state.count}</button>
    </section>
  );
};

export const LexicalScope = component$(() => {
  const state = useStore({
    count: 0,
  });
  const nu = 1;
  const str = 'hola';
  const obj = {
    a: { thing: 12 },
    b: 'hola',
    c: 123,
    d: false,
    e: true,
    f: null,
    g: undefined,
    h: [1, 'string', false, { hola: 1 }, ['hello']],
  };
  const noserialize = noSerialize({ text: 'not included', window: () => {} });
  const undef = undefined;
  const nulll = null;
  const array = [1, 2, 'hola', {}];
  const boolTrue = true;
  const boolFalse = false;
  const qrl = $(() => logDebug('qrl'));
  const el = useHostElement();
  const doc = useDocument();
  const thing = runtimeQrl(LexicalScope_render, [
    nu,
    str,
    obj,
    undef,
    nulll,
    array,
    boolTrue,
    boolFalse,
    state,
    noserialize,
    qrl,
    el,
    doc,
  ]);
  return <div onClick$={thing}></div>;
});
