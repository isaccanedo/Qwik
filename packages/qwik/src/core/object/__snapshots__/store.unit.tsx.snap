// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store should serialize content 1`] = `
Object {
  "ctx": Object {
    "#1": Object {
      "r": "0 1 2 9 a b d e f! 9 g #0 h",
    },
  },
  "objs": Array [
    1,
    "hola",
    Object {
      "a": "3",
      "b": "1",
      "c": 123,
      "d": "e",
      "e": "d",
      "f": "a",
      "g": "9",
      "h": "4",
    },
    Object {
      "thing": 12,
    },
    Array [
      "0",
      "5",
      "e",
      "6",
      "7",
    ],
    "string",
    Object {
      "hola": "0",
    },
    Array [
      "8",
    ],
    "hello",
    "",
    null,
    Array [
      "0",
      2,
      "1",
      "c",
    ],
    Object {},
    true,
    false,
    Object {
      "count": 0,
    },
    "/runtimeQRL#s1",
    "",
  ],
  "subs": Array [],
}
`;
